### PR TITLE
Update setHashOnion to limit the number of hashes

### LINKED ISSUE #8468: Hash onion computation can exhaust node resources
### Expected behavior
There should be limit on number of hash computation so that `memory` and `cpu` usage is bounded.
This RPC endpoints should be only enabled per whitelist with [this issue](https://github.com/LiskHQ/lisk-sdk/issues/8410). Therefore, this endpoint should be only called from the node operator by default.

### Actual behavior
With the very high number of count for the `setHashOnion`  like below.

### Steps to reproduce
Call RPC endpoint with
```
{"jsonrpc": "2.0", "id": 1, "method": "random_setHashOnion", "params":{"address":
"lskkgfk84gngotqjsqfujkctshvkwzs8z8w93gmja", "count": 1.7976931348623157e+308}}
```

### Which version(s) does this affect? (Environment, OS, etc...)
6.0.0-beta.1

### LINKED ISSUE #8469: setHashOnion RPC method does not validate seed length
### Expected behavior
Input to the setHashOnion endpoint should be validated

### Actual behavior
https://github.com/LiskHQ/lisk-sdk/blob/89e7504ef5eb6183aefe576a93be3d6052e56038/framework/src/modules/random/schemas.ts#L27-L50

Length of the seed is not validated. Therefore, it can use longer or shorter bytes as the seed which does not align with the [LIP](https://github.com/LiskHQ/lips/blob/3acef71ed72a521de73e4e173932490c9eacbd20/proposals/lip-0022.md#b-hash-onion-computation)

### Steps to reproduce
Call setHashOnion with other length than 16 bytes seed

### Which version(s) does this affect? (Environment, OS, etc...)
6.0.0-beta.1

### PR DESCRIPTION
**Pull request by shuse2**
### What was the problem?

This PR resolves #8468 and resolves #8469

### How was it solved?

- Add capability to directly set `hashes`
- Limit number of hash computation on a node to 10000
- Add validation for seed and hashes to be 16 bytes

### How was it tested?

- Added test cases for all the scenarios
- Check the RPC with multiple scenarios like below
```
    "jsonrpc": "2.0",
    "id": "1",
    "method": "random_getHashOnionSeeds",
    "params": {
        "address": "lskzbqjmwmd32sx8ya56saa4gk7tkco953btm24t8",
        "count": 10000,
        "distance": 1
    }
}
```

### CODE REVIEW

**Review by mosmartin — APPROVED** [review]

File: framework/test/unit/modules/random/endpoint.spec.ts [hunk 1]
```diff
@@ -255,6 +295,123 @@ describe('RandomModuleEndpoint', () => {
...
+			// Arrange
+			const { address } = genesisValidators.validators[0];
+			const count = 1000000;
+			context.params = { address, hashes: ['7c73f00f64fcebbab49a6145ef14a843'], count };
+
+			// Act & Assert
+			await expect(randomEndpoint.setHashOnion(context)).rejects.toThrow(
+				'Hashes must be provided with count and distance.',
+			);
+		});
```
Comment (bobanm) [inline framework/test/unit/modules/random/endpoint.spec.ts:401]: This test case is actually a copy of an existing test case `should throw error when hashes is provided but not distance`

File: framework/test/unit/modules/random/endpoint.spec.ts [hunk 2]
```diff
@@ -255,6 +295,123 @@ describe('RandomModuleEndpoint', () => {
...
+
+		it('should throw error when hashes empty', async () => {
```
Comment (bobanm) [inline framework/test/unit/modules/random/endpoint.spec.ts:336]:
```suggestion
		it('should throw error when hashes has an element not 16 bytes', async () => {
```

File: framework/test/unit/modules/random/endpoint.spec.ts [hunk 3]
```diff
@@ -255,6 +295,123 @@ describe('RandomModuleEndpoint', () => {
...
+			// Act & Assert
+			await expect(randomEndpoint.setHashOnion(context)).rejects.toThrow(
+				'must NOT have more than 32 characters',
+			);
+		});
+
+		it('should throw error when count is not multiple of distance', async () => {
+			// Arrange
+			const { address } = genesisValidators.validators[0];
+			const count = MAX_HASH_COMPUTATION * 10;
```
Comment (bobanm) [inline framework/test/unit/modules/random/endpoint.spec.ts:356]: To make the test focused, how about decreasing the count to `MAX_HASH_COMPUTATION`?

And then also adding a separate test which passes when `count > MAX_HASH_COMPUTATION` and hashes are provided?

File: framework/src/modules/random/endpoint.ts [hunk 4]
```diff
@@ -63,10 +63,36 @@ export class RandomEndpoint extends BaseEndpoint {
...
 			? Buffer.from(ctx.params.seed, 'hex')
 			: cryptography.utils.generateHashOnionSeed();
-		const count = ctx.params.count ?? 1000000;
+		const count = ctx.params.count ?? MAX_HASH_COMPUTATION;
 		const distance = ctx.params.distance ?? 1000;
 
-		const hashes = cryptography.utils.hashOnion(seed, count, distance) as Buffer[];
+		if (distance > count) {
+			throw new Error('Invalid distance. Distance must be less than or equal to count.');
+		}
```
Comment (bobanm) [inline framework/src/modules/random/endpoint.ts:71]: Is this needed? Isn't it always required that count must be a multiple of distance? Checking that should also cover this case, right? 🤔

File: framework/src/modules/random/endpoint.ts [hunk 5]
```diff
@@ -63,10 +63,36 @@ export class RandomEndpoint extends BaseEndpoint {
...
+
+		let hashes: Buffer[];
+		if (ctx.params.hashes) {
+			// when hashes is provided, count and distance must be provided together
+			if (!ctx.params.count || !ctx.params.distance) {
+				throw new Error('Hashes must be provided with count and distance.');
+			}
+			if (count % distance !== 0) {
+				throw new Error('Invalid count. Count must be multiple of distance.');
+			}
```
Comment (bobanm) [inline framework/src/modules/random/endpoint.ts:81]: Isn't this check relevant also when the hashes are not provided, and should go before `if (ctx.params.hashes)` check?

**Review by bobanm — CHANGES_REQUESTED** [review]

File: framework/test/unit/modules/random/endpoint.spec.ts (hunk 2 above)
Comment (shuse2) [reply framework/test/unit/modules/random/endpoint.spec.ts:336]: rephrased the error message. the test case was switched with something else, and `empty` was correct to check the empty array

File: framework/src/modules/random/endpoint.ts (hunk 5 above)
Comment (shuse2) [reply framework/src/modules/random/endpoint.ts:81]: yes, good catch 👍

**Pull request merged**